File: data-commons/src/main/java/se/tink/android/repository/service/UserConfigurationServiceCachedImpl.java
package se.tink.android.repository.service;

import com.google.common.collect.Lists;
import com.tink.model.user.UserProfile;
import com.tink.service.handler.ResultHandler;
import com.tink.service.observer.ChangeObserver;
import com.tink.service.user.UserProfileService;
import java.util.List;
import javax.inject.Inject;

//TODO: Core setup - Reevaluate
public class UserConfigurationServiceCachedImpl implements UserConfigurationService {

	private List<ChangeObserver<UserProfile>> changeObservers;
	private UserProfileService userService;

	@Inject
	public UserConfigurationServiceCachedImpl(UserProfileService userService) {
		this.userService = userService;

		changeObservers = Lists.newArrayList();
	}

	@Override
	public void subscribe(ChangeObserver<UserProfile> listener) {
		changeObservers.add(listener);
	}

	@Override
	public void unsubscribe(ChangeObserver<UserProfile> listener) {
		changeObservers.remove(listener);
	}

	@Override
	public void refreshUserConfiguration() {
		userService.getProfile(new ResultHandler<>(
			this::notifyOnRead,
				error -> {
					//TODO("Core setup")
				}
			)
		);
	}

	private void notifyOnRead(UserProfile userProfile) {
		for (ChangeObserver<UserProfile> observer : changeObservers) {
			observer.onRead(userProfile);
		}
	}
}

File: data-commons/src/main/java/se/tink/android/repository/service/UserConfigurationServiceCachedImpl.kt
package se.tink.android.repository.service

import com.tink.model.user.UserProfile
import com.tink.service.observer.ChangeObserver
import com.tink.service.user.UserProfileService
import com.tink.service.util.DispatcherProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

@Deprecated("Use UserProfileService directly")
class UserConfigurationServiceCachedImpl @Inject constructor(
    private val userService: UserProfileService,
    private val dispatcher: DispatcherProvider,
) : UserConfigurationService {

    private val changeObservers: MutableList<ChangeObserver<UserProfile>> = mutableListOf()

    override fun subscribe(listener: ChangeObserver<UserProfile>) {
        changeObservers.add(listener)
    }

    override fun unsubscribe(listener: ChangeObserver<UserProfile>) {
        changeObservers.remove(listener)
    }

    override fun refreshUserConfiguration() {
        CoroutineScope(dispatcher.io()).launch {
            try {
                notifyOnRead(userService.getProfile())
            } catch (error: Throwable) {
                Timber.e(error)
            }
        }
    }

    private fun notifyOnRead(userProfile: UserProfile) {
        for (observer in changeObservers) {
            observer.onRead(userProfile)
        }
    }
}